packages/errors/lib/base-error.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * @typedef {object} ErrorStrictData
+ * @property {string} [code]
+ *     A machine-readable error code which identifies the specific type of error.
+ * @property {string} [message]
+ *     A human readable message which describes the error.
+ * @property {Error | null} [cause]
+ *     The root cause error instance.
+ */
+
+/**
+ * @typedef {ErrorStrictData & Record<string, any>} ErrorData
+ */
+
+/**
+ * Class representing an error.
+ */
+class BaseError extends Error {
+	/**
+	 * @override
+	 * @readonly
+	 * @public
+	 * @type {string}
+	 */
+	name = 'BaseError';
+
+	/**
+	 * Whether the error is operational.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {boolean}
+	 */
+	isOperational = false;
+
+	/**
+	 * A machine-readable error code which identifies the specific type of error.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {string}
+	 */
+	code = BaseError.defaultCode;
+
+	/**
+	 * The root cause error instance.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {Error | null}
+	 */
+	cause = null;
+
+	/**
+	 * Additional error information.
+	 *
+	 * @readonly
+	 * @public
+	 * @type {{[key: string]: any}}
+	 */
+	data = {};
+
+	/**
+	 * Create an error with no arguments.
+	 *
+	 * @overload
+	 */
+	/**
+	 * Create an error with error data.
+	 *
+	 * @overload
+	 * @param {ErrorData} data
+	 *     Additional error information.
+	 */
+	/**
+	 * Create an error with a message and optional error data.
+	 *
+	 * @overload
+	 * @param {string} message
+	 *     The error message.
+	 * @param {ErrorData} [data]
+	 *     Additional error information.
+	 */
+	/**
+	 * @param {string | ErrorData} [message]
+	 *     The error message if it's a string, or full error information if an object.
+	 * @param {ErrorData} [data]
+	 *     Additional error information if `message` is a string.
+	 */
+	constructor(message, data = {}) {
+		if (typeof message === 'string') {
+			data.message = message;
+		} else {
+			data = message || data;
+		}
+		super(data.message || BaseError.defaultMessage);
+
+		if (typeof data.code === 'string') {
+			this.code = BaseError.normalizeErrorCode(data.code);
+		}
+
+		if (data.cause instanceof Error) {
+			this.cause = data.cause;
+		}
+
+		for (const [key, value] of Object.entries(data)) {
+			// @ts-ignore TypeScript does not properly infer the constructor
+			if (!this.constructor.reservedKeys.includes(key)) {
```

Comment:
**thought**: is there a way to do typecasting with JSDoc? in Typescript i'd do `(this.constructor as typeof BaseError)`

Reply:
OMG it is! I'm gonna go and update all of the other ts-ignores for this issue [0afacc8](https://github.com/Financial-Times/dotcom-reliability-kit/pull/648/commits/0afacc8f631c0b58cc4cd1eb4b9995d65c65cb81)